src/Middleware/Authentication.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Auth\Middleware;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Auth\AuthenticationMethodInterface;
+use Yiisoft\Auth\Handler\AuthenticationFailureHandler;
+use Yiisoft\Strings\StringHelper;
+
+/**
+ * Authentication middleware tries to authenticate and identity using request data.
+ * If identity is found, it is set to request attribute allowing further middleware to obtain and use it.
+ * If identity is not found failure handler is called. By default it is {@see AuthenticationFailureHandler}.
+ */
+final class Authentication implements MiddlewareInterface
+{
+    private AuthenticationMethodInterface $authenticationMethod;
+
+    /**
+     * @var RequestHandlerInterface A handler that is called when there is a failure authenticating an identity.
+     */
+    private RequestHandlerInterface $failureHandler;
+
+    /**
+     * @var array Patterns to match to consider the given request URI path optional.
+     */
+    private array $optional = [];
+
+    public function __construct(
+        AuthenticationMethodInterface $authenticationMethod,
+        ResponseFactoryInterface $responseFactory,
+        RequestHandlerInterface $authenticationFailureHandler = null
+    ) {
+        $this->authenticationMethod = $authenticationMethod;
+        $this->failureHandler = $authenticationFailureHandler ?? new AuthenticationFailureHandler(
+            $responseFactory
+        );
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $identity = $this->authenticationMethod->authenticate($request);
+        $request = $request->withAttribute(self::class, $identity);
+
+        if ($identity === null && !$this->isOptional($request)) {
+            return $this->authenticationMethod->challenge(
+                $this->failureHandler->handle($request)
+            );
+        }
+
+        return $handler->handle($request);
+    }
+
+    /**
+     * @param array $optional Patterns to match to consider the given request URI path optional.
+     * @return self
+     */
+    public function withOptional(array $optional): self
```

Comment:
How about adding a method that combines the previous `optional`.

Reply:
I don't see a use-case for it.